const with nested result type gives c compiler error: lvalue expected

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
V full version: V 0.4.8 18eee34.04aba8f
OS: windows, Microsoft Windows 11 Pro v22631 64-bit
Processor: 24 cpus, 64bit, little endian, 

getwd: D:\vbug2
vexe: C:\v\v.exe
vexe mtime: 2024-09-30 05:47:23

vroot: OK, value: C:\v
VMODULES: OK, value: C:\Users\info\.vmodules
VTMP: OK, value: C:\Users\info\AppData\Local\Temp\v_0

Git version: git version 2.42.0.windows.2
Git vroot status: weekly.2024.39-50-g04aba8f7-dirty
.git/config present: true

CC version: Error: 'cc' is not recognized as an internal or external command,

operable program or batch file.


thirdparty/tcc status: thirdparty-windows-amd64 b425ac82

```

**What did you do?**
`./v -g -o vdbg cmd/v && ./vdbg main.v`
```v
module main

struct Foo {
	bar int
}

__global (
	a Foo
)

const c = [[a.bar]!]!

fn main () {}
```

**What did you expect to see?**

no c error


**What did you see instead?**
```
/c/v/v -enable-globals .
================== C compilation error (from tcc): ==============
cc: C:/Users/info/AppData/Local/Temp/v_0/vbug2.01J915FGCEVMZTY3M7XEGNXYY1.tmp.c:7019: warning: implicit declaration of function 'tcc_backtrace'
cc: C:/Users/info/AppData/Local/Temp/v_0/vbug2.01J915FGCEVMZTY3M7XEGNXYY1.tmp.c:13510: error: lvalue expected
... (the original output was 3 lines long, and was truncated to 2 lines)
=================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
builder error:
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```
> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -6247,7 +6247,13 @@ fn (mut g Gen) const_decl_init_later_msvc_string_fixed_array(mod string, name st
 	cname := g.c_const_name(name)
 	mut init := strings.new_builder(100)
 	for i, elem_expr in expr.exprs {
-		init.writeln(g.expr_string_surround('\t${cname}[${i}] = ', elem_expr, ';'))
+		if elem_expr is ast.ArrayInit {
+			elem_typ := g.typ(elem_expr.typ)
+			init.writeln(g.expr_string_surround('\tmemcpy(${cname}[${i}], (${elem_typ})',
+				elem_expr, ', sizeof(${elem_typ}));'))
+		} else {
+			init.writeln(g.expr_string_surround('\t${cname}[${i}] = ', elem_expr, ';'))
+		}
 	}
 	mut def := '${styp} ${cname}'
 	g.global_const_defs[util.no_dots(name)] = GlobalConstDef{

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -6247,7 +6247,7 @@ fn (mut g Gen) const_decl_init_later_msvc_string_fixed_array(mod string, name st
 	cname := g.c_const_name(name)
 	mut init := strings.new_builder(100)
 	for i, elem_expr in expr.exprs {
-		if elem_expr is ast.ArrayInit {
+		if elem_expr is ast.ArrayInit && elem_expr.is_fixed {
 			elem_typ := g.typ(elem_expr.typ)
 			init.writeln(g.expr_string_surround('\tmemcpy(${cname}[${i}], (${elem_typ})',
 				elem_expr, ', sizeof(${elem_typ}));'))

--- a/vlib/v/tests/consts/const_array_fixed_test.v
+++ b/vlib/v/tests/consts/const_array_fixed_test.v
@@ -10,7 +10,23 @@ __global (
 )
 
 const c = [[a.bar]!]!
+const d = [[a.bar]]!
+const e = [[a.bar]!]
+const f = [[[a.bar]!]!]!
+const g = [a.bar]
 
-fn test_main() {
-	dump(c == [[120]!]!)
+fn test_selector() {
+	assert dump(c == [[a.bar]!]!)
+	assert dump(d == [[a.bar]]!)
+	assert dump(e == [[a.bar]!])
+	assert dump(f == [[[a.bar]!]!]!)
+	assert dump(g == [a.bar])
+}
+
+fn test_literal() {
+	assert dump(c == [[120]!]!)
+	assert dump(d == [[120]]!)
+	assert dump(e == [[120]!])
+	assert dump(f == [[[120]!]!]!)
+	assert dump(g == [120])
 }

--- a/vlib/v/tests/consts/const_array_fixed_test.v
+++ b/vlib/v/tests/consts/const_array_fixed_test.v
@@ -1,0 +1,16 @@
+@[has_globals]
+module main
+
+struct Foo {
+	bar int = 120
+}
+
+__global (
+	a Foo
+)
+
+const c = [[a.bar]!]!
+
+fn test_main() {
+	dump(c == [[120]!]!)
+}

